cpp/onedal/detail/table_impl_iface.hpp

```diff
@@ -0,0 +1,36 @@
+/*******************************************************************************
+ * Copyright 2020 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#pragma once
+
+#include "onedal/detail/storage_iface.hpp"
+#include "onedal/table_metadata.hpp"
+
+namespace dal::detail {
+
+class table_impl_iface : public dense_storage_iface<storage_readable_writable> {
```

Comment:
**I see a tremendous problem with the selected approach and binary compatibility in the user-defined numeric table.** `table_impl_wrapper` implements the `table_impl_iface` which is subclass of `dense_storage_iface<...>` that provides methods for accessors. We designed accessors as separate classes to be able to add a new one quickly. 

The problem is that the proposed approach requires extension of `dense_storage_iface` with new virtual methods to support new new accessors. Adding new virtual methods to the `dense_storage_iface` (or adding to table_impl new ifaces to inherit from) changes table of virtual functions (extends, even if we assume they are added in order).

Let's image the following situation -- case of dynamic binary compatibility where we provide new binaries with old headers:
- We added new accessor in new library version, to be able to work with the new accessor, we've extended storage iface with the new virtual method (it's all in header files);
- The new binary assumes that the virtual table is extended, let's say it stores 10 pointers to functions;
- The old user's code (their table) compiled with the **old** `dense_storage_iface` ensures only 8 pointers in vtable;
- When the user's object is passed to new binary, the new code in our binary tries to call those newly-added virtual methods. So it leads to crash...

Reply:
This is very important problem, I've been thinking about it all day, written a couple of prototypes. The solution for that problem is not simple we can meet tomorrow and discuss approaches. I would formulate the problem as more general -- how to create a binary stable abstract class. In particular, how to add new virtual functions to the class keeping binary compatibility.